compiler/aarch64/codegen/BinaryEvaluator.cpp

```diff
@@ -723,6 +723,138 @@ OMR::ARM64::TreeEvaluator::lremEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+         {
+         sourceNode = andNode->getSecondChild();
+         maskNode = andNode->getFirstChild();
+         }
+      }
+
+   if (maskNode == NULL)
+      {
+      return NULL;
+      }
+   const bool is64bit = shiftNode->getDataType().isInt64();
+   const int64_t shiftValue = shiftValueNode->getConstValue();
+
+   if ((shiftValue <= 0) || (shiftValue > (is64bit ? 63 : 31)))
+      {
+      return NULL;
+      }
+   const uint64_t maskValue = is64bit ? maskNode->getLongInt() : static_cast<uint32_t>(maskNode->getInt());
+
+   if (shiftNode->getOpCode().isLeftShift())
+      {
+      if ((maskValue << shiftValue) == 0)
+         {
+         /*
+          * The result is always 0 in this case.
+          */
+         TR::Register *reg = cg->allocateRegister();
+         generateTrg1ImmInstruction(cg, TR::InstOpCode::movzx, shiftNode, reg, 0);
+         shiftNode->setRegister(reg);
+         cg->recursivelyDecReferenceCount(andNode);
+         cg->decReferenceCount(shiftValueNode);
+         return reg;
+         }
+
+      if (((maskValue & 1) == 1) && (((maskValue >> (is64bit ? 63 : 31)) & 1) == 0) && contiguousBits(maskValue))
+         {
+         /*
+          * maskValue has consecutive 1s from the least significant bits. The most significant bit must be 0. Otherwise, mask is all 1s.
+          * So, this mask and shift operation is copying consecutive bits starting from the least significant bits of the source register
+          * to bit position shiftValue of the destination register.
+          */
+         uint32_t width = populationCount(maskValue);
+         uint32_t imms = (width - 1);
+         uint32_t immr = (is64bit ? 64 : 32) - shiftValue;
+
+         TR::Register *reg = cg->allocateRegister();
+         TR::Register *sreg = cg->evaluate(sourceNode);
+
+         generateTrg1Src1ImmInstruction(cg, (is64bit ? TR::InstOpCode::ubfmx : TR::InstOpCode::ubfmw), shiftNode, reg, sreg, (immr << 6) | imms);
+         shiftNode->setRegister(reg);
+         cg->recursivelyDecReferenceCount(andNode);
+         cg->decReferenceCount(shiftValueNode);
+         return reg;
+         }
+      }
+   else /* right shift*/
+      {
+      uint64_t shiftedMask = (maskValue >> shiftValue);
+      /*
+       * If the lsb of shiftedMask is set and the shiftedMask has consecutive 1s, then this operation is copying 
+       * consecutive 1s starting from the bit position shiftValue of the source register to the least significant bits of the destination register.
+       */
+      if (((shiftedMask & 1) == 1) && contiguousBits(shiftedMask))
+         {
+         uint32_t width = populationCount(shiftedMask);
+         uint32_t shiftRemainderWidth = (is64bit ? 64 : 32) - shiftValue;
+
+         bool isRightShift = (width == shiftRemainderWidth);  /* In this case, it works as a shift. */
+
+         uint32_t imms = (shiftValue + width - 1);
+         uint32_t immr = shiftValue;
+         TR::Register *reg = cg->allocateRegister();
```

Comment:
The same comment for the reference count of `sourceNode`.

Reply:
force-pushed the updated commit, too.